check for known-slave in sentinel rewrite config

jodevsa opened the pull request:
closes the following issue: https://github.com/redis/redis/issues/11774


This PR attempts to fix the following config file error

```
*** FATAL CONFIG FILE ERROR (Redis 6.2.7) ***
Reading the configuration file, at line 152
>>> 'sentinel known-replica XXXX 127.0.0.1 5001'
Duplicate hostname and port for replica.
```


that is happening when a user uses the legacy key "known-slave" in the config file and a config rewrite occurs. The config rewrite logic won't replace the old  line "sentinel known-slave XXXX 127.0.0.1 5001" and would add a new line with "sentinel known-replica XXXX 127.0.0.1 5001" which results in the error above "Duplicate hostname and port for replica."


example:

Current sentinal.conf
```
...

sentinel known-slave XXXX 127.0.0.1 5001
sentinel example-random-option X
...
```
after the config rewrite logic runs:
```
....
sentinel known-slave XXXX 127.0.0.1 5001
sentinel example-random-option X

# Generated by CONFIG REWRITE
sentinel known-replica XXXX 127.0.0.1 5001
```











This bug only exists in Redis versions >=6.2 because prior to that it was hidden by the effects of this bug https://github.com/redis/redis/issues/5388 that that @hwware fixed in https://github.com/redis/redis/pull/8271  and was released in versions >=6.2

moticless commented:
Hi @jodevsa,
Please write good description to the PR. It is not sufficient to give a link. Regarding the implementation, you set a  hook for a sentinel related string at core infrastructure (config.c). This is clearly not the way to do it. Please take the time to learn the code. 

Consider assist @hwware, I saw that he suggested his  help.

Thank you

jodevsa commented:
Hi @moticless, 

Thank you for your feedback :) This is still a draft and still up for modifications.

`Regarding the implementation, you set a hook for a sentinel related string at core infrastructure`
I understand that. based on my understanding I think core infrastructure should still handle this case otherwise we might get the same problem years later if someone forgot to handle "known-slave" together with "known-replica". 

@moticless @hwware I'm still new to this project and I would really appreciate any recommendation on how this should be implemented.

hwware commented:
I suggest you could update your PR title to "WIP --- check for known-slave in sentinel rewrite config" (working in process status), so that everyone can see your PR status, and then add good description in the top comment area. (As moti said)
Finally, becuase this is a sentinel issue, you could find some rewrite codes in the src/sentinel.c

Any problem is welcome, good luck ^_^

jodevsa commented:
Thanks @hwware. btw, can you confirm that this is an issue? I assumed that when I saw that you opened an issue for it but never got an explicit confirmation from you that this is something that needs to be fixed ( sorry for assuming things)

hwware commented:
I think this is a redis legacy issue. 
If you check redis old version, before redis 4.0.14, including 4.0.14, there is no known-replica, only known-slave is there.
Since redis 5, known-replica appears and it is encouraged to use it, but client could still choose known-slave.

Thus, my suggestion is:
If in the sentinel profile file, there is known-replica parameter or there is no this parameter, when client run the config rewrite command, we could save to the known-replica format.

However, if client define known-slave parameter in the profile, we should save to the  known-slave parameter format if config rewrite command is run.

jodevsa commented:
> if there is known-replica parameter or there is no known-replica and known-slave parameter,
How can I read that information from inside `sentinel.c`? I don't see any function to do that?

jodevsa commented:
description added and I changed my implementation. @hwware  can you give me your opinion?

jodevsa commented on src/config.h:
```diff
@@ -27,6 +27,18 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
+struct rewriteConfigState {
```
Note to myself: move after ifndef

jodevsa commented on src/sentinel.c:
```diff
@@ -2263,7 +2279,7 @@ void rewriteConfigSentinelOption(struct rewriteConfigState *state) {
     rewriteConfigMarkAsProcessed(state,"sentinel config-epoch");
     rewriteConfigMarkAsProcessed(state,"sentinel leader-epoch");
-    rewriteConfigMarkAsProcessed(state,"sentinel known-replica");
+    rewriteConfigMarkAsProcessed(state,knownReplicaConfig);
```
open question - what happens if we have both "known-replica" and "known-slave" in the file?

hwware replied:
I think the config parameters are pre-config by human, thus generally known-replica and known-slave should not exist together.

hwware commented on src/config.h:
```diff
@@ -30,6 +30,18 @@
+    int force_write;      /* True if we want all keywords to be force
+                             written. Currently only used for testing
+                             and debug information. */
+};
```
Why add this definition in this PR? I do not understand

jodevsa replied:
This was originally in config.c. i needed to move it out to the header file so that I can access that struct from sentinel.c

hwware commented on src/sentinel.c:
```diff
@@ -2055,6 +2056,20 @@ void rewriteConfigSentinelOption(struct rewriteConfigState *state) {
 
+    /* detirmine if the config uses the legacy known-slave or known-replica term */
+    const char *knownReplicaConfig;
+
```
maybe we do not need this space

jodevsa replied:
Is there any utility to auto fix format?

hwware commented on src/sentinel.c:
```diff
@@ -2055,6 +2056,20 @@ void rewriteConfigSentinelOption(struct rewriteConfigState *state) {
+    sds o = sdsnew("sentinel known-slave");
+    list *l = dictFetchValue(state->option_to_line,o);
+    
+    if(l){
```
it should include one space between ( and {

hwware commented on src/sentinel.c:
```diff
@@ -2055,6 +2056,20 @@ void rewriteConfigSentinelOption(struct rewriteConfigState *state) {
+    if(l){
+        knownReplicaConfig="sentinel known-slave";
+    }
+    else{
```
else should be align with the }

moticless commented:
Hi @jodevsa,
IMHO, your changes exposes too much internals of config.c unnecessarily.
I think with a slight API modification of `rewriteConfigRewriteLine()` it can
indicates whether it managed to rewrite or not (return 1 iff rewrite) and if it
failed non-forced-rewrite of `known-slave`, then force-rewrite `known-replica`.

So the modified code at sentinel would look like this:
```
  if (rewriteConfigRewriteLine(state,"sentinel known-slave",sdsdup(line),0) == 0) {
    rewriteConfigRewriteLine(state, "sentinel known-replica", line, 1);
  } else {
    sdsfree(line);
 }
```
Of course it need to be reviewed and  fully tested that don't miss anything.

jodevsa commented:
Hi,

Thanks!! I feel bad for not figurig out this easy way 😅

jodevsa commented:
Hey @moticless @hwware ,

I implemented the required changes.

moticless commented on src/config.c:
```diff
@@ -1220,7 +1220,7 @@ struct rewriteConfigState *rewriteConfigReadOldFile(char *path) {
  * "line" is either used, or freed, so the caller does not need to free it
  * in any way. */
-void rewriteConfigRewriteLine(struct rewriteConfigState *state, const char *option, sds line, int force) {
+int rewriteConfigRewriteLine(struct rewriteConfigState *state, const char *option, sds line, int force) {
```
I think we should return 1 if-and-only-if we managed to replace a line.
This way the caller will be able to distinct between the case a new line
was appended due to force flag (last arg) vs. a line was actually replaced.

jodevsa replied:
Hey, somehow didn't see this comment. Can you please give me more information on why do we need that distinction between a replace case and a new line case? I thought we only need to know that the line was not replaced when using force=false.

So are you suggesting something like

return 0 // line not replaced or added
return 1 // line replaced
return 2 // line added

?

moticless commented on src/sentinel.c:
```diff
@@ -2175,7 +2175,11 @@ void rewriteConfigSentinelOption(struct rewriteConfigState *state) {
                 "sentinel known-replica %s %s %d",
                 master->name, announceSentinelAddr(slave_addr), slave_addr->port);
-            rewriteConfigRewriteLine(state,"sentinel known-replica",line,1);
+            /* try to replace any known-slave option first if found */
+            if(rewriteConfigRewriteLine(state,"sentinel known-slave",line,0) ==0){
+                rewriteConfigRewriteLine(state, "sentinel known-replica", line, 1);
+            }
+            sdsfree(line);
```
I think you are having double free of memory. Function `rewriteConfigRewriteLine()`
is the one to release `line`. I think you should do instead:
```
  if (rewriteConfigRewriteLine(state,"sentinel known-slave",sdsdup(line),0) == 0) {
    rewriteConfigRewriteLine(state, "sentinel known-replica", line, 1);
  } else {
    sdsfree(line);
 }
```
I didn't suggest to write automated test. The least we can do is to test it 
manually throughly that it is working as expected. Please verify that:
1. if having a configuration file with "known-slave" then the configuration file 
    after update will only include "known-replica".
3.  And if having a configuration with "known-replica" then configuration file
   after update is having expected data with only "known-replica".

Thank you

jodevsa replied:
Thank you @moticless :)

hwware commented:
Any update for this PR? Thanks

jodevsa commented:
<WIP>

Hi, I've implemented the required changes and did the following testing:



**testing the bug on the `unstable` branch:**

1- > `make`
2- create sentinel.conf with the following contents:
```
port 26379
daemonize no
sentinel monitor mymaster 127.0.0.1 6379 2
SENTINEL known-slave mymaster 127.0.0.1 5001
```
3- > `./src/redis-sentinel sentinel.conf`

output:
```
31216:X 19 Mar 2023 21:12:47.965 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
31216:X 19 Mar 2023 21:12:47.965 # Redis version=255.255.255, bits=64, commit=e7f35edb, modified=0, pid=31216, just started
31216:X 19 Mar 2023 21:12:47.965 # Configuration loaded
31216:X 19 Mar 2023 21:12:47.966 * Increased maximum number of open files to 10032 (it was originally set to 256).
31216:X 19 Mar 2023 21:12:47.966 * monotonic clock: POSIX clock_gettime
                _._
           _.-``__ ''-._
      _.-``    `.  `_.  ''-._           Redis 255.255.255 (e7f35edb/0) 64 bit
  .-`` .-```.  ```\/    _.,_ ''-._
 (    '      ,       .-`  | `,    )     Running in sentinel mode
 |`-._`-...-` __...-.``-._|'` _.-'|     Port: 26379
 |    `-._   `._    /     _.-'    |     PID: 31216
  `-._    `-._  `-./  _.-'    _.-'
 |`-._`-._    `-.__.-'    _.-'_.-'|
 |    `-._`-._        _.-'_.-'    |           https://redis.io
  `-._    `-._`-.__.-'_.-'    _.-'
 |`-._`-._    `-.__.-'    _.-'_.-'|
 |    `-._`-._        _.-'_.-'    |
  `-._    `-._`-.__.-'_.-'    _.-'
      `-._    `-.__.-'    _.-'
          `-._        _.-'
              `-.__.-'

31216:X 19 Mar 2023 21:12:47.977 # WARNING: The TCP backlog setting of 511 cannot be enforced because kern.ipc.somaxconn is set to the lower value of 128.
31216:X 19 Mar 2023 21:12:47.998 * Sentinel new configuration saved on disk
31216:X 19 Mar 2023 21:12:47.998 # Sentinel ID is e7ecadbb739b53b12f12bf8ca3a64d7ca9ea4875
31216:X 19 Mar 2023 21:12:47.998 # +monitor master mymaster 127.0.0.1 6379 quorum 2
31216:X 19 Mar 2023 21:13:18.003 # +sdown master mymaster 127.0.0.1 6379
31216:X 19 Mar 2023 21:13:18.003 # +sdown slave 127.0.0.1:5001 127.0.0.1 5001 @ mymaster 127.0.0.1 6379
^[[A

```


4- stop redis-sentinel


5- examine `sentinel.conf` content

```
port 26379
daemonize no
sentinel monitor mymaster 127.0.0.1 6379 2
SENTINEL known-slave mymaster 127.0.0.1 5001
# Generated by CONFIG REWRITE
dir "/Users/subhialhasan/projects/redis"
protected-mode no
latency-tracking-info-percentiles 50 99 99.9
user default on nopass sanitize-payload ~* &* +@all
sentinel myid 5e3eb1658d685364a07d21eaa863590c871c6b4a
sentinel config-epoch mymaster 0
sentinel leader-epoch mymaster 0
sentinel known-replica mymaster 127.0.0.1 5001
sentinel current-epoch 0

```
we can notice that the file contains both `known-replica` and `known-slave` which is what this PR is trying to fix

jodevsa commented:
Hi @moticless and @hwware, sorry for the delays :( 

I've implemented the changes required and included the testing done in the previous message.

hwware commented:
The codes look good to me. When you have chance, please take a review and merge it.  Thanks @oranagra @madolson

jodevsa commented:
thanks!

oranagra commented:
sentinel CI: https://github.com/redis/redis/actions/runs/4605723009
